airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -0,0 +1,269 @@
+        else:
+            raise RuntimeError("Global substream cursor timer not started")
+
+
+class GlobalSubstreamCursor(DeclarativeCursor):
+    """
+    The GlobalSubstreamCursor is designed to track the state of substreams using a single global cursor.
+    This class is beneficial for streams with many partitions, as it allows the state to be managed globally
+    instead of per partition, simplifying state management and reducing the size of state messages.
+
+    This cursor is activated by setting the `global_substream_cursor` parameter for incremental sync.
+
+    Warnings:
+    - This class enforces a minimal lookback window for substream based on the duration of the previous sync to avoid losing records. This lookback ensures that any records added or updated during the sync are captured in subsequent syncs.
+    - The global cursor is updated only at the end of the sync. If the sync ends prematurely (e.g., due to an exception), the state will not be updated.
+    - When using the `incremental_dependency` option, the sync will progress through parent records, preventing the sync from getting infinitely stuck. However, it is crucial to understand the requirements for both the `global_substream_cursor` and `incremental_dependency` options to avoid data loss.
+    """
+
+    def __init__(self, stream_cursor: DeclarativeCursor, partition_router: PartitionRouter):
+        self._stream_cursor = stream_cursor
+        self._partition_router = partition_router
+        self._timer = Timer()
+        self._lock = threading.Lock()
+        self._slice_semaphore = threading.Semaphore(0)  # Start with 0, indicating no slices being tracked
+        self._all_slices_yielded = False
+        self._lookback_window: Optional[int] = None
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        """
+        Generate stream slices and ensure that the last slice is flagged appropriately.
+        This method uses a semaphore to track the processing of slices and ensures
+        that `close_slice` is called only after all slices have been processed.
+        """
+        previous_slice = None
+
+        slice_generator = (
+            StreamSlice(partition=partition, cursor_slice=cursor_slice)
+            for partition in self._partition_router.stream_slices()
+            for cursor_slice in self._stream_cursor.stream_slices()
+        )
+
+        for slice in slice_generator:
+            if previous_slice is not None:
+                # Release the semaphore to indicate that a slice has been yielded
+                self._slice_semaphore.release()
+                yield previous_slice
+
+            # Store the current slice as the previous slice for the next iteration
+            previous_slice = slice
+
+        # After all slices have been generated, release the semaphore one final time
+        # and flag that all slices have been yielded
+        self._slice_semaphore.release()
```

Comment:
I like that we have added a comment, but I'm still not sure it is sufficient to understand why we release. Should we be more explicit about why? It feels like it has something to do about not releasing on the first slice (see `if previous_slice is not None` at line 70) but it is not clear to me

Reply:
The semaphore is released every time before yielding the slice to increase the counter by one, indicating one more slice to process.

Reply:
Ok. I think by brain fart is that we set `self._all_slices_yielded = True` before we have actually yielded all the slices but the fact that we release the semaphore just before prevents `close_slice` to actually close the slice. Can we expand on the comment? This might be overkill but it feels like there are a lot of decision/sequencing of operation that would be worth making explicit. Something like:

```
We expect the following events:
* Yields all the slices except the last one. At this point, `close_slice` won't actually close the global slice as `self._all_slices_yielded == False`
* Release the semaphore one last time before setting `self._all_slices_yielded = True`. This will cause `close_slice` to know about all the slices before we indicate that all slices have been yielded so the left side of `if self._all_slices_yielded and self._slice_semaphore._value == 0` will be false if not everything is closed
* Setting `self._all_slices_yielded = True`. We do that before actually yielding the last slice as the caller of `stream_slices` might stop iterating at any point and hence the code after `yield` might not be executed
* Yield the last slice. At that point, once there are as many slices yielded as closes, the global slice will be closed too
```

Reply:
Thanks, great comment. I will add it to the docstring